### DISCUSSION TITLE
Web (PWA) vs Nativo: Qual escolher?

### DISCUSSION THREAD

**Discussion by kazzkiq** [frontendbr/forum#470] [Perguntas e Respostas]
Fala galera,

Essa é uma dúvida recorrente que eu tenho e imagino que outras pessoas aqui também se perguntem: **Hoje em dia, ainda vale a pena focar em aplicativos "nativos"?**

Com a W3C trazendo APIs cada vez mais avançadas para web, e com as *engines* dos navegadores cada vez mais performáticas, será que vale a pena focar em criar "aplicativos nativos" (embarcados em WebViews com Cordova e derivados) ao invés de adotar novas metodologias como [Progressive Web Apps](https://www.smashingmagazine.com/2016/08/a-beginners-guide-to-progressive-web-apps/)?

E do ponto de vista de adoção do usuário, será que um app nativo ainda ganha mais visibilidade (e gera mais acessos) do que um [web app bem feito](https://pwa.rocks/)? **Afinal, usuários preferem acessar um app via URL ou via App Store?**

Se você tiver opiniões ou alguma experiência pra compartilhar sobre essa *polêmica*, por favor, não hesite em comentar!

**Comment by felipefialho** [comment]
Invoco o @sergiolopes :)

E linko essa issue #34

**Comment by matheusml** [comment]
Oi @kazzkiq, vamos a sua pergunta:

> Hoje em dia, ainda vale a pena focar em aplicativos "nativos"?

Na minha visão, a resposta certa infelizmente é: **depende**.

#### Cenário 1:
Você é uma empresa pequena, ou mesmo uma _startup_, seu _budget_ é bem apertado e você tem um número pequeno de usuários, então talvez pro seu caso a construção de um _app_ nativo seja um investimento caro demais.

Pra esse cenário específico, talvez o ideal seja construir um Progressive Web App e medir os resultados. E com base nesses resultados decidir se o investimento em um _app_ nativo possa valer a pena.

#### Cenário 2:
Você é uma empresa média/grande, com um bom _budget_ para investir e você já possui uma quantidade razoável de usuários, então na minha visão é um cenário quase que perfeito para a construção de um _app_ nativo.

Só tem uma observação: se esse for o seu caso, nada te impede também de construir um PWA.
O custo dos PWAs em relação aos _apps_ nativos é baixíssimo, então talvez seja o caso de construir ambos, medir o resultado e depois investir mais naquele que teve melhores retornos.

#### Para saber mais
É sempre importante também estudar a fundo as tecnologias antes de querer implementar.
Vou recomendar alguns _links_, tanto em inglês quanto em português, que possam ajudar a entender melhor os PWAs:

- [Introdução aos Progressive Web Apps](https://medium.com/tableless/introdu%C3%A7%C3%A3o-aos-progressive-web-apps-ad47ba24cddb#.abjcuxzup)
- [Google | Progressive Web Apps](https://developers.google.com/web/progressive-web-apps/)
- [Palestra do Sergio Lopes no Front In Sampa 2015](https://www.youtube.com/watch?v=sH7dlRnuh-k)

Abraços!

**Comment by sergiolopes** [comment]
Tirando o argumento do tempo/custo de construção de lado por um momento, **o que é melhor pro usuário?** Depende bastante. 

A moda das Apps passou já. [O usuário médio instala zero Apps por mês.](https://qz.com/253618/most-smartphone-users-download-zero-apps-per-month/) O tempo médio de uso de Apps é alto, mas metade do tempo é no Facebook e o resto no top 4 Apps (WhatsApp, Instagram, Waze etc). É muito mais fácil alcançar usuários na Web, a não ser que vc seja uma marca grande e tenha um apelo **muito** forte pra alguém instalar sua App (se estiver na dúvida é porque você não tem um apelo muito forte).

Existem muitos casos de quem saiu de App e foi pra Web e alcançou mais usuários ([Flipkart é o mais famoso](https://developers.google.com/web/showcase/2016/flipkart)).

A principal diferença é que clicar num link é muito mais fácil e natural que 1) buscar na Store; 2) clicar em instalar; 3) aceitar as permissoes; 4) esperar instalar; 5) abrir a App. A Web favorece o uso pontual e descompromissado, por isso captura mais usuários. Claro, o modelo de App fideliza mais ([ou não](http://www.androidauthority.com/77-percent-users-dont-use-an-app-after-three-days-678107/)) com icone na home, push etc, mas é justo isso que as PWAs tentam trazer. Uma experiência essencialmente Web mas que pode ser usada pra fidelizar a (pequena) parcela de usuários que querem aquele algo a mais além de um link.

Mas tudo isso não quer dizer que lançar na Store não faça sentido. Existe um bom número de usuários que usa sim várias Apps, instala novas e busca primeiro na Store do que no Google. E ter seu ícone lá vai atender esse público. O importante é saber que esse público é menor que o público da Web aberta. Mas ele existe e você pode querer atacá-lo.

No fundo quando discutimos Web vs Apps, estamos discutindo **mídia**. E sua marca quer estar onde o público está. O público maior é na Web, mas há o público fanboy de App. Assim como tem o público Facebook - que experencia a Internet dentro do Facebook e procura Fanpages antes de Googlar qualquer coisa.

Então, a não ser que tenha alguma limitação técnica, se você puder, esteja em todos os lugares: Web, App Stores, Facebook, chatbots etc. 

Mas aí voltamos o argumento principal: **custo e priorização**. Você não consegue atacar todas as mídias do mundo a não ser que seja gigante. E mesmo os gigantes, analisam a presença potencial do público dele em certa mídia antes de investir (ex. Google não tem Apps pra Windows Phone até hoje).

É bem difícil saber o que priorizar, eu sei. Mas pra 90% (inventei o número) dos cenários, eu enxergo uma estratégia Web-first como a mais correta. E se fizer uma PWA bem feita, é bem fácil empacotá-la pra distribuir nas lojas tbm - pelo menos até as lojas da Apple e Google começarem a aceitar PWAs (a Windows Store já permite).

E, claro, sempre há o caso particular, do cenário particular, que nada disso que falei faz sentido. Cada caso é um caso e eu no fundo só tô botando mais lenha na fogueira :)

**Comment by mksanti** [comment]
> pelo menos até as lojas da Apple e Google começarem a aceitar PWAs (a Windows Store já permite).

que legal.... nem sabia que a Windows Store permitia, darei olhada.

**Comment by WandersonAlves** [comment]
>  E se fizer uma PWA bem feita, é bem fácil empacotá-la pra distribuir nas lojas tbm 

Eu acredito que PWA é um rumo até melhor do que precisar de uma das lojas para distribuição. Como que começa a distribuição de uma aplicação (minha opnião).

- Caso publica: começa no site da empresa, com a divulgação para as pessoas acessarem esse site para conhecerem, a partir disso, o usuário pode começar a usar o PWA (e pode até existir uma apelação do tipo "hey, você não precisa instalar pra usar"). Bem marketing
- Caso privado: não teria muita divulgação, inserir na loja seria lucro, seria algo distribuído "de boca", a cada um novo install, o usuário pode ou não compartilhar e falar "hey, app bacana, procura lá". Porém usuário nenhum irá dar essa garantia de compartilhamento.

Uma PWA bem feita pode sim descartar o uso de uma aplicação nativa (minha opinião hehe).

Clássico user case: https://developers.google.com/web/showcase/2016/flipkart

**Comment by johanguse** [comment]
Good News povo! https://www.windowscentral.com/great-equalizer-microsoft-google-and-apple-embrace-progressive-web-apps

Já esta implementado no Safari https://webkit.org/blog/8060/release-notes-for-safari-technology-preview-47/

**Comment by giovannism20** [comment]
Sim, @johanguse desde Agosto...

**Comment by jlameira** [comment]
Pessoal, 
Queria saber sobre a segurança em aplicações PWA.
Tipo os App's na hora da instalação te perguntam sobre a permissão de acesso de alguns recursos do celular,  tanto que na hora que se publica o App, vc tem que informar todos os recursos que o App vai usar do celular.
Como se dá esse cenário de segurança de acesso a recursos do celular nos PWA's, pois pelo que entendi eles não são instaláveis e sim uma Web Browser.

**Comment by felipemfp** [comment]
@jlameira, pelo que entendi do que já vi, você vai ter acesso às APIs que são fornecidas pelos browsers. Então essa questão de pedir permissão, o próprio celular sobe um popup quando você pedir o microfone, por exemplo.

**Comment by ngoctrai1997** [comment]
existem muitas webs famosas agora usam o Pwa como uma maneira diferente de alcançar seus clientes como o Facebook, Instagram, então eu acho que o Pwa vai crescer mais
visite esta web para entender mais sobre o pwa [https://tapita.io/](url)

**Comment by 0xjoaovictor** [comment]
> Tirando o argumento do tempo/custo de construção de lado por um momento, **o que é melhor pro usuário?** Depende bastante.
> 
> A moda das Apps passou já. [O usuário médio instala zero Apps por mês.](https://qz.com/253618/most-smartphone-users-download-zero-apps-per-month/) O tempo médio de uso de Apps é alto, mas metade do tempo é no Facebook e o resto no top 4 Apps (WhatsApp, Instagram, Waze etc). É muito mais fácil alcançar usuários na Web, a não ser que vc seja uma marca grande e tenha um apelo **muito** forte pra alguém instalar sua App (se estiver na dúvida é porque você não tem um apelo muito forte).
> 
> Existem muitos casos de quem saiu de App e foi pra Web e alcançou mais usuários ([Flipkart é o mais famoso](https://developers.google.com/web/showcase/2016/flipkart)).
> 
> A principal diferença é que clicar num link é muito mais fácil e natural que 1) buscar na Store; 2) clicar em instalar; 3) aceitar as permissoes; 4) esperar instalar; 5) abrir a App. A Web favorece o uso pontual e descompromissado, por isso captura mais usuários. Claro, o modelo de App fideliza mais ([ou não](http://www.androidauthority.com/77-percent-users-dont-use-an-app-after-three-days-678107/)) com icone na home, push etc, mas é justo isso que as PWAs tentam trazer. Uma experiência essencialmente Web mas que pode ser usada pra fidelizar a (pequena) parcela de usuários que querem aquele algo a mais além de um link.
> 
> Mas tudo isso não quer dizer que lançar na Store não faça sentido. Existe um bom número de usuários que usa sim várias Apps, instala novas e busca primeiro na Store do que no Google. E ter seu ícone lá vai atender esse público. O importante é saber que esse público é menor que o público da Web aberta. Mas ele existe e você pode querer atacá-lo.
> 
> No fundo quando discutimos Web vs Apps, estamos discutindo **mídia**. E sua marca quer estar onde o público está. O público maior é na Web, mas há o público fanboy de App. Assim como tem o público Facebook - que experencia a Internet dentro do Facebook e procura Fanpages antes de Googlar qualquer coisa.
> 
> Então, a não ser que tenha alguma limitação técnica, se você puder, esteja em todos os lugares: Web, App Stores, Facebook, chatbots etc.
> 
> Mas aí voltamos o argumento principal: **custo e priorização**. Você não consegue atacar todas as mídias do mundo a não ser que seja gigante. E mesmo os gigantes, analisam a presença potencial do público dele em certa mídia antes de investir (ex. Google não tem Apps pra Windows Phone até hoje).
> 
> É bem difícil saber o que priorizar, eu sei. Mas pra 90% (inventei o número) dos cenários, eu enxergo uma estratégia Web-first como a mais correta. E se fizer uma PWA bem feita, é bem fácil empacotá-la pra distribuir nas lojas tbm - pelo menos até as lojas da Apple e Google começarem a aceitar PWAs (a Windows Store já permite).
> 
> E, claro, sempre há o caso particular, do cenário particular, que nada disso que falei faz sentido. Cada caso é um caso e eu no fundo só tô botando mais lenha na fogueira :)

Passando aqui para mostrar que o que foi dito a 2 anos atrás, está acontecendo:
[Google Play Store now open for PWA](https://medium.com/@firt/google-play-store-now-open-for-progressive-web-apps-ec6f3c6ff3cc)
